pkg/loadbalancer/flow/nfthandler.go

```diff
@@ -0,0 +1,475 @@
+	"github.com/nordix/meridio/pkg/loadbalancer/types"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+/* Example config:
+table inet meridio-nfqlb {
+
+        # (things omitted...)
+
+        set vips4-flow-port-nat {
+                type ipv4_addr
+                flags interval
+                elements = { 10.0.0.1 }
+        }
+
+        set vips6-flow-port-nat {
+                type ipv6_addr
+                flags interval
+                elements = { 1000::1:a00:1 }
+        }
+
+        chain flow-port-nat {
+                type filter hook prerouting priority 100; policy accept;
+                ip daddr @vips4-flow-port-nat tcp dport 7777 tcp dport set 5001 counter packets 305 bytes 16348 notrack
+                ip saddr @vips4-flow-port-nat tcp sport 5001 tcp sport set 7777 counter packets 180 bytes 11472 notrack
+                ip6 daddr @vips6-flow-port-nat tcp dport 7777 tcp dport set 5001 counter packets 600 bytes 44160 notrack
+                ip6 saddr @vips6-flow-port-nat tcp sport 5001 tcp sport set 7777 counter packets 209 bytes 17496 notrack
+        }
+}
+
+# Manual config;
+nft list ruleset
+tbl=meridio-nfqlb
+chain=flow-port-nat
+nft add chain inet $tbl $chain {type filter hook prerouting priority 100\;}
+nft flush chain inet $tbl $chain
+nft --debug all add rule inet $tbl $chain ip daddr @ipv4-vips tcp dport 7777 tcp dport set 5001 counter notrack
+nft --debug all add rule inet $tbl $chain ip saddr @ipv4-vips tcp sport 5001 tcp sport set 7777 counter notrack
+
+# The output from "nft --debug ..." is used to write the buildRule functions.
+*/
+
+// Sync with "github.com/nordix/meridio/pkg/loadbalancer/nfqlb"
+// or export.
+const (
+	tableName = "meridio-nfqlb"
+)
+
+type nfth struct {
+	table *nftables.Table
+}
+
+// NewNftHandler returns an NFT handler used for port-NAT rules
+func NewNftHandler() (types.NftHandler, error) {
+	conn := &nftables.Conn{}
+
+	// Get the nft table
+	tables, err := conn.ListTables()
+	if err != nil {
+		logrus.Errorf("NewNftHandler:ListTables:%v", err)
+		return nil, err
+	}
+	var nfth nfth
+	for _, t := range tables {
+		if t.Name == tableName {
+			nfth.table = t
+			break
+		}
+	}
+	if nfth.table == nil {
+		err = fmt.Errorf("Table not found: %s", tableName)
+		logrus.Errorf("NewNftHandler:ListTables:%v", err)
+		return nil, err
+	}
+
+	return &nfth, nil
+}
+
+func (nfth *nfth) PortNATSet(
+	flowName string, protocols []string, dport, localPort uint) error {
+	if nfth.table == nil {
+		err := fmt.Errorf("Table not found: %s", tableName)
+		logrus.Errorf("PortNATSet:%v", err)
+		return err
+	}
+
+	conn := &nftables.Conn{}
+	ch := conn.AddChain(&nftables.Chain{
+		Name:     "flow-" + flowName,
+		Table:    nfth.table,
+		Type:     nftables.ChainTypeFilter,
+		Hooknum:  nftables.ChainHookPrerouting,
+		Priority: nftables.ChainPriorityNATSource,
+	})
+	conn.FlushChain(ch)
+
+	ipv4Vips, ipv6Vips := nfth.vipSets(flowName)
+	for _, prot := range protocols {
```

Comment:
Actually it is possible to use a single rule to match multiple l4 protocols port info to lower the overall number of rules (https://www.mankier.com/8/nft#Payload_Expressions-Raw_Payload_Expression). However I guess it does not seem plausible that a "nat flow" will be configured with multiple protocols, so it would not worth the effort.

Reply:
I let is be as-is. SCTP is not mentioned in the man page and may require special treatement anyway.